py-rse/py-scripting.Rmd

```diff
@@ -212,99 +212,222 @@ optional arguments:
+
+NAME
+    script_template - One-line description of what the script does.
+
+FUNCTIONS
+    main(args)
+        Run the program.
+
+FILE
+    /Users/amira/script_template.py
+```
+
+As this example shows,
+if we ask for help on the module,
+Python formats and displays all of the docstrings for everything in the file.
+
+## Counting Words {#py-rse-py-scripting-wordcount}
+
+Now that we have a template for command-line Python programs,
+we can use it to check Zipf's Law for our collection of classic English novels.
+We start by moving the template into the directory
+where we store various programs and shell scripts we have created
+so that we can refer to it later for other projects:
+
+FIXME: where do we introduce `~/bin`?
+
+```shell
+$ mv script_template.py ~/bin
+```
+
+Next,
+let's write a function that counts how often words appear in a file.
+Our function splits the text on [whitespace][whitespace] characters
+(which is the default behavior of the string object's `split` method),
+then strips leading and trailing punctuation.
+This isn't completely correct---if two words are joined by a long dash
+like "correct" and "if" in this sentence, for example,
+they will be treated as one word---so
+we will explore better options in the exercises.
+We also use the `Counter` class from the `collections` library
+to count how many times each word occurs.
+If we give `Counter` a list of words,
+the result is an object that contains
+the number of times each one appears in the list:
 
 ```python
-import re
+import string
 from collections import Counter
 
 
 def count_words(reader):
     """Count the occurrence of each word in a string."""
     text = reader.read()
-    findwords = re.compile(r"\w+", re.IGNORECASE)
-    word_list = re.findall(findwords, text)
+    chunks = text.split()
+    stripped = [word.strip(string.punctuation) for word in chunks]
+    word_list = [word for word in stripped if word]
     word_counts = Counter(word_list)
     return word_counts
 ```
 
-`\w+` means to match all characters that commonly occur inside words,
-such as letters, numbers, and underscores.
-Spaces and punctuation are not matched,
-so most strings will be split at these characters,
-effectively separating the text into individual words.
-
-We can apply that function to the Dracula ebook (for example):
+Let's try our function on *Dracula*:
 
 ```python
-with open('../data/dracula.txt', 'r') as reader:
+with open('data/dracula.txt', 'r') as reader:
     word_counts = count_words(reader)
 print(word_counts)
 ```
 
 ```text
-Counter({'the': 7474, 'and': 5803, 'I': 4846, 'to': 4662, 'of': 3707, 'a': 2955, 'in': 2466, 'that': 2436, 'he': 1996, 'was': 1870, 'it': 1808, 'is': 1498, 'for': 1480, 'as': 1476, 'me': 1452, ...
+Counter({'the': 7474, 'and': 5803, 'I': 4846, 'to': 4662,
+         'of': 3707, 'a': 2955, 'in': 2466, 'that': 2436,
+         'he': 1996, 'was': 1870, 'it': 1808, 'is': 1498,
+         'for': 1480, 'as': 1476, 'me': 1452, ...})
```

Comment:
The numbers change now that you're using string operators instead of regular expressions. 

Here's the new numbers with the addition of `word.lower()` from my previous comment:
```
Counter({'the': 8036, 'and': 5896, 'i': 4712, 'to': 4540, 'of': 3738, 'a': 2961, 'in': 2558, 'he': 2543, 'that': 2455, 'it': 2141, 'was': 1877, 'as': 1581, 'we': 1535, 'for': 1534, 'is': 1526, 'his': 1467, 'you': 1450, 'me': 1446,...
```
and here it is without the addition of `word.lower()`:
```
Counter({'the': 7430, 'and': 5723, 'I': 4712, 'to': 4489, 'of': 3697, 'a': 2908, 'in': 2457, 'that': 2397, 'he': 1961, 'was': 1865, 'it': 1781, 'is': 1498, 'as': 1458, 'for': 1452, 'me': 1443, 'his': 1381, 'not': 1374, 'with': 1271,...
```

Reply:
:+1: